Serious karengraphqlapi refactoring

Rembane opened the pull request:
So, this is the result of an insane amount of refactoring I did in january and february to simplify the code involved in fetching data from kåren's GraphQL API. Especially to make it use more data types and helper functions used in the other parts of the program. 

Then I made a pull request ( #76 ) which I and Jassob worked through thoroughly and closed when I found out that I had started implementing another way of fetching the current day from the API. That part of the code has now been removed and what is left is some quite nice Haskell if I may say so myself.

Rembane committed:
Removed deepLookup from Model.KarenGraphQLApi.

Rembane committed:
Removed the unit field from Meal. It's not used.

Rembane committed:
Only request the GraphQL data we use in Model.KarenGraphQLApi.

I removed all the other fields from the request.

You can specify what data you want when you make a GraphQL query, this
is definitely one of the killer features of GraphQL, and removing things
that are not used is neat too.

Rembane committed:
Replaced all occurences of nameOf with its body.

Rembane committed:
Replaced Mealname with Model.Types.Menu

* Inlined the only occurence of parseResponse.
* Removed Meal in KarenGraphQLApi.
  Replaced it with Menu from Model.Types. Replaced the FromJSON instance
  with a parseMenu function.

Rembane committed:
Refactored the fetch functions of KarenGraphQLApi.

They now have a distinct fetch stage and a parse stage which should make
debugging much easier.

@adamse found redundant parameters in menuParser during code review,
this is fixed in this commit. Thank you Adam!

Also Haddock style comments were added for the parameters of
fetchAndCreateRestaurant, and there was much rejoicing.

Rembane committed:
Added a function to evaluate our monad transformer stack.

Very useful for development and debugging purposes.

Rembane commented:
I rebased master on this branch and force pushed, that's why everything looks weird.

adamse commented on src/Model/KarenGraphQLApi.hs:
```diff
@@ -73,135 +72,93 @@ graphQLQuery
 
--- | Parameters: the date to fetch, title, tag, uuid
+  createMenuParser :: Language -> Value -> Parser Menu
+  createMenuParser l = withObject "Menu Object" $ \obj ->
```
Kan skippa `l` har o anvanda `lang` nedan. Kanske ocksa kalla den `menuParser`.

Rembane replied:
@adamse Ja, det stämmer bara bra i den commiten. Gäller det även i den senaste versionen av koden?

adamse replied:
Ser onekligen så ut, om inte github lurar mig så tittar jag på samtliga ändringar i PRet...

Rembane replied:
Jag tror Github är aslurig just nu på ett eller annat sätt, för såhär ser koden ut för mig på den senaste committen av den här branchen: 

```haskell
-- | Parses menus from Kåren's GraphQL API.
parse
  :: Language             -- ^ Language
  -> BL8.ByteString       -- ^ Bytestring payload from fetch
  -> Either NoMenu [Menu] -- ^ Either list of parsed Menu's or NoMenu error
parse lang =
  failWithNoMenu eitherDecode
    >=> failWithNoMenu
          (parseEither
            (   withObject "Parse meals"
            $   (.: "data")
            >=> (.: "dishOccurrencesByTimeRange")
            >=> mapM (createMenuParser lang)
            )
          )
    >=> menusToEitherNoLunch
 where
  failWithNoMenu :: Show a => (a -> Either String b) -> a -> Either NoMenu b
  failWithNoMenu action x =
    first (\msg -> NMParseError msg . BL8.pack . show $ x) (action x)

  createMenuParser :: Language -> Value -> Parser Menu
  createMenuParser l = withObject "Menu Object" $ \obj ->
    Menu
      <$> (obj .: "dishType" >>= (.: "name"))
      <*> ((obj .: "displayNames") >>= withArray
            "An array of meal names"
            (   mapM
                ( withObject "The name of the meal in many languages"
                $ \o' -> (,) <$> (o' .: "categoryName") <*> (o' .: "name")
                )
            >=> ( maybe (fail $ "Couldn't find the language: " <> show l)
                        (pure . snd)
                . find ((== l) . fst)
                )
            )
          )
```

Koden som finns där, är den okej, eller vill du att jag ska ändra något på den?

Hämtat härifrån: https://github.com/dtekcth/mat-chalmers/blob/serious-karengraphqlapi-refactoring/src/Model/KarenGraphQLApi.hs#L131

adamse replied:
```
x lang = ... y lang ...
  where y l = ... l ...
```

vs

```
x lang = ... y ...
  where y = ... lang ...
```

Rembane replied:
@adamse Nu förstod jag äntligen vad du menade. Jag har haft -3 INT de senaste dagarna pga förkylning och det märktes här. :D 

Jag har applicerat dina föreslagna förändringar här: ad72861.

Rembane replied:
Men den här är nog mer relevant just nu: 1f00947a

Rebase och force push är kul grejer.

Jassob commented:
I seemed to have made changes in my first commit, only to undo them in the next. If you let me, I'll rebase the intermediate state away

Rembane commented:
@Jassob sure, do it!

Jassob commented:
@Rembane I would like us to not have fixup-commits, but rather to go back and fix the commits themselves (although I do appreciate the way you give kudos :slightly_smiling_face:) 

I'd be happy to edit the history if you think that I'm unnecessarily pedantic :octocat:

Rembane commented:
@Jassob That's a very good point. I fix!

Rembane commented:
@Jassob Done, this got a bit out of hand, so there are fewer commits now than before. They just disappeared! It wasn't my fault! :O

Rembane committed:
Used Day instead of String for representing dates.

This way we get some type safety and it's much harder to represent
invalid data.

Rembane commented:
After thinking about it for a night I added a distinct commit (4ed6200) to replace `String` with `Day`. Not really necessary, but makes the series of commits a bit clearer.

Rembane commented:
Thank you @adamse and @Jassob for the review and help!
